src/capture.cpp

```diff
@@ -386,8 +395,48 @@ void acquisition::Capture::read_parameters() {
             distort_list_provided = true;
         }
     }
+    
+    bool rect_list_provided = false;
+    XmlRpc::XmlRpcValue rect_list;
+
+    if (nh_pvt_.getParam("rectification_coeffs", rect_list)) {
+        ROS_INFO("  Camera Rectification Paramters:");
+        ROS_ASSERT_MSG(rect_list.size() == num_ids,"If rectification_coeffs are provided, they should be the same number as cam_ids and should correspond in order!");
+        for (int i=0; i<rect_list.size(); i++){
+            std::vector<double> rect;
+            String rect_str="";
+            for (int j=0; j<rect_list[i].size(); j++){
+                ROS_ASSERT_MSG(rect_list[i][j].getType()== XmlRpc::XmlRpcValue::TypeDouble,"Make sure all numbers are entered as doubles eg. 0.0 or 1.1");
+                rect.push_back(static_cast<double>(rect_list[i][j]));
+                rect_str = rect_str +to_string(rect[j])+" ";
+            }
+            rect_coeff_vec_.push_back(rect);
+            ROS_INFO_STREAM("   "<< rect_str );
+            rect_list_provided = true;
+        }
+    }
+    
+    bool proj_list_provided = false;
+    XmlRpc::XmlRpcValue proj_list;
+
+    if (nh_pvt_.getParam("projection_coeffs", proj_list)) {
+        ROS_INFO("  Camera Projection Paramters:");
+        ROS_ASSERT_MSG(proj_list.size() == num_ids,"If projection_coeffs are provided, they should be the same number as cam_ids and should correspond in order!");
+        for (int i=0; i<proj_list.size(); i++){
+            std::vector<double> proj;
+            String proj_str="";
+            for (int j=0; j<proj_list[i].size(); j++){
+                ROS_ASSERT_MSG(proj_list[i][j].getType()== XmlRpc::XmlRpcValue::TypeDouble,"Make sure all numbers are entered as doubles eg. 0.0 or 1.1");
+                proj.push_back(static_cast<double>(proj_list[i][j]));
+                proj_str = proj_str +to_string(proj[j])+" ";
+            }
+            proj_coeff_vec_.push_back(proj);
+            ROS_INFO_STREAM("   "<< proj_str );
+            proj_list_provided = true;
+        }
+    }
 
-    PUBLISH_CAM_INFO_ = intrinsics_list_provided && distort_list_provided;
+    PUBLISH_CAM_INFO_ = intrinsics_list_provided && distort_list_provided && proj_list_provided && rect_list_provided;
```

Comment:
We had written this node to be generic with multiple cameras  and not just 2.  I don't believe that a rectification and projection matrix is well defined when you have more than 2 cameras. (Which baseline would you choose etc.) Thus this condition of all 4 having to be provided doesn't make sense when you have more than 2 cameras.
In these cases it still makes sense to have intrinsic and distortion_coefficients specified, so that you could use the camera location from a URDF to add features to the images in say rVIZ.
A better way to handle this might be:
- If intrinsics and distortion coeffs provided correctly publish them
- If exactly 2 cameras provided and proj_matrics and rect_matrics provided correctly, additionally publish that too.

Thoughts?

Reply:
I am no expert on camera calibration matrices, but I thought that every camera will have a separate list of matrices (projection, rectification, intrinsic and distortion). So, if you have three cameras, you will have 3 projection matrix, 3 distortion matrix etc. What I do not understand is that, if each camera is going to have a separate list of matrices, then how does number of cameras have any affect on these? What I am understanding from you comment is that cameras are going to somehow share the projection and rectification matrix [please correct me if I misunderstood your comment].

I think we should only publish camera_info if intrinsic and distortion values are provided. And optionally publish projection, and rectification values if they are provided.

Reply:
From the description of CameraInfo messages:
```
# Rectification matrix (stereo cameras only)
# A rotation matrix aligning the camera coordinate system to the ideal
# stereo image plane so that epipolar lines in both stereo images are
# parallel.
float64[9]  R # 3x3 row-major matrix

# Projection/camera matrix
#     [fx'  0  cx' Tx]
# P = [ 0  fy' cy' Ty]
#     [ 0   0   1   0]
# By convention, this matrix specifies the intrinsic (camera) matrix
#  of the processed (rectified) image. That is, the left 3x3 portion
#  is the normal camera intrinsic matrix for the rectified image.
# It projects 3D points in the camera coordinate frame to 2D pixel
#  coordinates using the focal lengths (fx', fy') and principal point
#  (cx', cy') - these may differ from the values in K.
# For monocular cameras, Tx = Ty = 0. Normally, monocular cameras will
#  also have R = the identity and P[1:3,1:3] = K.
# For a stereo pair, the fourth column [Tx Ty 0]' is related to the
#  position of the optical center of the second camera in the first
#  camera's frame. We assume Tz = 0 so both cameras are in the same
#  stereo image plane. The first camera always has Tx = Ty = 0. For
#  the right (second) camera of a horizontal stereo pair, Ty = 0 and
#  Tx = -fx' * B, where B is the baseline between the cameras.
# Given a 3D point [X Y Z]', the projection (x, y) of the point onto
#  the rectified image is given by:
#  [u v w]' = P * [X Y Z 1]'
#         x = u / w
#         y = v / w
#  This holds for both images of a stereo pair.
float64[12] P # 3x4 row-major matrix
```
We could leave it with your proposed solutions of publishing the projection and rectification values if provided.

Reply:
Done.